pymc3/tests/test_distributions_random.py

```diff
@@ -492,43 +421,316 @@ class TestMoyal(BaseTestCases.BaseTestCase):
     params = {"mu": 0.0, "sigma": 1.0}
 
 
-@pytest.mark.skip(reason="This test is covered by Aesara")
-class TestCategorical(BaseTestCases.BaseTestCase):
-    distribution = pm.Categorical
-    params = {"p": np.ones(BaseTestCases.BaseTestCase.shape)}
+class TestRandomVariableDistribution(SeededTest):
```

Comment:
1. I realized the test above `BaseTestCases` is using `pm.Model()` and `pm.Distribution(name, *args, **kwargs)` instead of `pm.Distribution.dist()`. That sounds like a better approach as that is **actually** how users will interact with distributions. The `.dist()` syntax is rarely used and may actually be dropped in the future. I am sorry for nudging you towards `.dist()`, do you think the tests could work equally well with the other approach?

https://github.com/pymc-devs/pymc3/blob/e4901eeada712cd1f64189a6acc01fcab84f1fc2/pymc3/tests/test_distributions_random.py#L148

2. Do you think we should merge the other two types of tests currently in this file within this general framework? I am talking about the test mentioned in point 1. above which explicitly deals with shapes, and the two older tests `pymc3_random`/ `pymc3_random_discrete` that perform a statistical comparison between a large number of samples of a pymc3 and a reference distribution, to see if they match. The shape test will be important for all ops implemented on our side, while the later two will only be rarely used for distributions that don't have a scipy / numpy direct counterpart but can still be compared to transformations / specific parametrizations of other random distributions (see https://github.com/pymc-devs/pymc3/pull/4596 for an example)

3. I am not sure whether the syntax with the dictionaries and list of tests is the best, but I see why you went with it.

Reply:
Alternatively these tests can live separately from the other two types of tests, in which case I would rename this class of tests to something like `test_pymc3_randomOp_parametrization`

Reply:
1. ok
2. yes, that's the idea. I can add the one dealing with shapes for the gumbel distribution (so that it can be used as a reference). Regarding `pymc3_random` / `pymc3_random_discrete` tests, I guess that the outcome the test is trying to achieve should be the same of `_check_pymc_dist_matches_expected`, so possibly this should be changed to handle non-exact comparisons? after this change the user would specify in `params` whether it's not an exact comparison and whether it's a discrete or continuous distribution and then the right way to compare the two distributions will be picked accordingly (ie current `_check_pymc_dist_matches_expected` or `pymc3_random` or `pymc3_random_discrete`)? wdyt?
3. `I am not sure whether the syntax with the dictionaries and list of tests is the best` me neither, I don't find it particularly elegant tbh. Still, it allows for having only one test per distribution and leaves the flexibility to specify what checks to run. I couldn't think of any other design that had these two benefits and didn't end up being very verbose, or required to draw samples (or run eval()) multiple times. I think that this approach is probably only acceptable for tests (little need for modularity/reusability) Happy to replace it with a better one, if you already have one in mind

Reply:
3. a more pleasant-to-read approach would be creating a base class similar to `BaseTestCase` that defines all the tests and utils needed. Subclass for each distribution, define in each subclass the appropriate parameters individually as class variables (similar to `BaseTestCase`) and have a list of callables containing the tests to ran (same as `tests_to_run` now) as another class variable.

Reply:
>  pymc3_random / pymc3_random_discrete tests, I guess that the outcome the test is trying to achieve should be the same of _check_pymc_dist_matches_expected, so possibly this should be changed to handle non-exact comparisons?

On a second thought, I wouldn't bother integrating them. We might actually not find much use for them. If they were to be integrated I would make them a separate tests from `_check_pymc_dist_matches_expected`, something like `check_draws_match_statistical_test` or something. In the past we were only using them to wastefully compare with the scipy functions directly or with the same random code in our distributions, both of which are better accomplished by your new exact tests.

Reply:
> yes, that's the idea. I can add the one dealing with shapes for the gumbel distribution (so that it can be used as a reference) 

Maybe add it to the `Beta` as we are using a custom `rng_fn`. The `Gumbel` is a direct aesara op so it's not the "best" example.

Reply:
>     1. a possibly more pleasant-to-read approach would be creating a base class similar to `BaseTestCase` that defines all the tests and utils needed. Subclass for each distribution, define in each subclass the appropriate parameters individually as class variables (similar to `BaseTestCase`) and have a list of callables containing the tests to ran (same as `tests_to_run` now) as another class variable. This is probably much nicer actually than the current one

Sounds good!

Reply:
> 1. The `.dist()` syntax is rarely used and may actually be dropped in the future.

Yes, and we should create an issue for that.  It should be a fairly simple change to make.

Also, we should do something about the mandatory name argument.  We definitely can and should remove the unique name requirement, but that's a separate issue.